text/trace/0107-sampling-score.md

```diff
@@ -0,0 +1,217 @@
+  also configured to use `ExternalScoreSampler`)
+  - `ExternalScoreSampler.ShouldSample` checks if score is provided in
+    `tracestate`.
+  - makes sampling decision by comparing upstream-generated score with its
+    sampling rate
+  - if span will be recorded: sampler adds `sampling.score` attribute to
+    attributes collection of to-be-created span
+
+Here is a [proof of concept](https://github.com/lmolkova/opentelemetry-dotnet/pull/1)
+in .NET.
+
+### Specification Delta
+
+1. Add `SamplingResult.Tracestate` field: sampler should be able to assign a
+   new tracestate for to-be-created span
+2. Add convention for `sampling.score` attribute on span (TBC). Check out
+   [open questions](open-questions) regarding attribute vs special field.
+3. Add `ExternalScoreSampler` implementation of `Sampler`
+
+### Trade-offs and mitigations
+
+This change would be the first (AFAIK) common use case of the `tracestate`.
+It comes with bandwidth and performance overhead: `tracestate` could have
+been just propagated [blindly](https://github.com/open-telemetry/opentelemetry-specification/issues/478).
+and the overhead is made before sampling decision and cannot be mitigated.
+
+Customers should configure it explicitly to avoid the overhead in the default
+case when interoperability is not necessary.
+
+Vendors may gradually update their existing solutions to support external
+score in order to interoperate with OpenTelemetry and should recommend
+customers to configure such sampler.
+
+It may be the case that after migration to OpenTelemetry is finalized, the need
+of `sampling.score` will decrease and customers can remove
+`ExternalScoreSampler` from configuration.
+
+## Prior art and alternatives
+
+Related discussions on [Probability sampler](https://github.com/open-telemetry/opentelemetry-specification/pull/570)
+
+### Sampling.score is NOT priority
+
+Priority is used by [OpenTracing](https://github.com/opentracing/specification/blob/master/semantic_conventions.md)
+as an implementation-specific hint for sampler to prioritize recording a span.
+
+[OpenTelemetry collector](https://github.com/open-telemetry/opentelemetry-collector/blob/60b03d0d2d503351501291b30836d2126487a741/processor/samplingprocessor/probabilisticsamplerprocessor/testdata/config.yaml#L10)
+uses `sampling.priority` to hint collector's sampler decision
+
+To avoid conflicts with existing implementations we fo not reuse priority term.
+
+## Open questions
+
+### Score calculation: can we use ProbabilitySampler?
+
+This spec suggests to generate score randomly to achieve uniform
+distribution.
+
+Assuming trace-ids are uniformly distributed, `ProbabilitySampler` can generate
+score, so the flow could look like this:
+
+`ExternalScoreSampler.ShouldSample`:
+
+- checks if `sampling.score` is available in the tracestate
+- if it's not there, invokes `ProbabilitySampler`, which calculates score
+  and populates it on the attributes
```

Comment:
I like this.

Reply:
so I've tried to follow this approach in the new version. To make it clean, I suggest separating score generation from 
sampling.

Samplers can use score generation, and also attach attributes, change tracestate.
If we allow falling back to another sampler:
- we need to do score propagation back from one sampler to the ExtenralScore one
- coordinate possible multiple changes of tracestate and attributes

If we have a layer responsible for score calculation (random, deterministic of any sort), we can make it much simpler and cleaner.